File: JavaScript/src/decorators/Aged.js
import ItemDecorator from './ItemDecorator';

// Increases quality of an item over time
export default class Aged extends ItemDecorator {
  static MAXIMUM_QUALITY = 50;

  tick() {
    this.daysRemaining -= 1;

    this.quality += this.#calculateIncrease();
    this.quality = Math.min(
      this.quality,
      this.constructor.MAXIMUM_QUALITY
    );
  }

  #calculateIncrease() {
    let increase = 1;
    if (this.daysRemaining <= 0) {
      increase += 1;
    }

    return increase;
  }
}

File: Ruby/lib/decorators/aged.rb
# frozen_string_literal: true

require 'delegate'

module Decorators
  # Increases quality of an item over time
  class Aged < SimpleDelegator
    MAXIMUM_QUALITY = 50

    alias item __getobj__

    def tick
      item.days_remaining -= 1

      item.quality += calculate_increase
      item.quality = [quality, MAXIMUM_QUALITY].min
    end

    private def calculate_increase
      increase = 1
      increase += 1 if days_remaining <= 0

      increase
    end
  end
end
